brainstorming.md

```diff
@@ -0,0 +1,372 @@
+
+## SOMADataFrame
+
+SOMADataFrame is a 1-D multi-column table of monomorphic arrays, all of equal length -- ie, a "dataframe". The SOMADataFrame has a user-defined schema, which includes:
+
+- a table schema, expressed as an `Arrow.RecordBatch` type, defining number of columns, and their respective names and types
+- the index column name
+
+A SOMADataFrame must contain at least a single column which is the named index column. The index column _must_ be of `string` type. Index values are non-null and must be unique (ie, no duplicates are allowed) within the object. The index is sparse and unordered, ie, supported indexing is by ID value only, with no indexing order or positional (offset-based) indexing.
+
+Most language-specific implementations will present SOMADataFrame in more convenient forms, such as Python `pandas.DataFrame`, R `data.frame` or other environment-specific data structures. Specific implementations of the SOMA API may extend this type with convertors to/from the underlying data structure, and enforce indexing conventions.
+
+> ⚠️ **Issue** - this spec requires that index column is a `string` type. Most underlying "engines" will support more flexibility. Do we want to enable that flexibility, or stick with requiring string for simplicity/commonality?
+
+## SOMAMatrix
+
+SOMAMatrix is a 2-D multi-column table of monomorphic arrays, ie, a "dataframe" with two index dimensions. SOMAMatrix is an extension of SOMADataFrame, with two indexing dimensions. The SOMAMatrix has a user-defined schema, which includes:
+
+- a table schema, expressed as an `Arrow.RecordBatch` type, defining the number of columns, and their respective names and types
+- the _primary_ index column name
+- the _secondary_ index column name
+
+A SOMAMatrix must contain (minimally) two columns which are the primary and secondary index columns. The index columns _must_ be of `string` type. Index values are non-null and must be unique (ie, no duplicates are allowed) within the object. The index is sparse and unordered, ie, supported indexing is by ID value only, with no indexing order or positional (offset-based) indexing.
+
+> ⚠️ **Issue** - SOMAMatrix and SOMADataFrame differ only in the number of dimensions (2 vs 1 respectively). Should they be unified into a single type object/API of more general nature? Clarity vs flexibility?
+
+## SOMA
+
+The SOMA type is a specialized SOMACollection, representing an annotated 2-D matrix of observations across a set of variables. The SOMA composes SOMACollection, SOMADataFrame and SOMAMatrix types using a set of predefined fields (keys & values), to represent the annotated matrix. Each predefined field in the SOMA has a well-defined type, dimensionality and indexing convention. Other user defined data may be added to a SOMA, as a SOMA is a specialization of a SOMACollection.
```

Comment:
here too -- keeping it loose/drafty -- just giving the reader an 'anchor' to hold onto while reading/absorbing/grokking -- mention we _might_ think of `SOMA` as "the thing" that has `obs`/`var`/`X` that single-cell savvy readers will see and head-nod at

Reply:
@johnkerl - could you clarify your comment?  I'm not understanding what the concern is...  Thanks!

Reply:
I will add some prose as you suggest.  I do not think your literal suggestion is what we want, as a new reader will have no idea what `soma.X` is or implies.  I'll state something in more standard English -- see if it is sufficient for a new reader...

Reply:
Prose added to the SOMACollection section.